pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,287 @@
+	parentNodePath := filepath.Dir(currNodePath)
+	numPAddrCells, numAddrSpaceCells, err := getNumAddrSizeCellsInfo(parentNodePath)
+	if err != nil {
+		return INVALID_ADDR, err
+	}
+
+	var addrRanges []rangeInfo
+	if physAddress == INVALID_ADDR { // Case 1: We are the Child Node. No addr has been set. Read the reg file to get the physical address within our parents space.
+		physAddress, err = getRegAddr(currNodePath, numPAddrCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+	} else { // Case 2: We use the ranges property to continue mapping a child addr into our parent addr space.
+		addrRanges, err = getRangesAddrInfo(currNodePath, numCAddrCells, numPAddrCells, numAddrSpaceCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+		// getRangesAddrInfo returns a list of all possible child address ranges our physical address can fall into. We must see which range to use, so that we can map our physical address into the correct parent address range.
+		for _, addrRange := range addrRanges {
+
+			if addrRange.childAddr <= physAddress && physAddress <= addrRange.childAddr+addrRange.addrSpaceSize {
+				physAddress -= addrRange.childAddr  // get the offset beween the address and child base address
+				physAddress += addrRange.parentAddr // now address has been mapped into parent space.
+				break
+			}
+		}
+	}
+
+	numCAddrCells = numPAddrCells
+	currNodePath = parentNodePath
+	return setGPIONodePhysAddrHelper(currNodePath, physAddress, numCAddrCells)
+}
+
+// Creates a virtual Page to access/manipulate memory related to gpiochip data
+func (b *pinctrlpi5) createGPIOVPage(memPath string) error {
+	var err error
+
+	// Open the 'file' you are trying to map. Note: /dev/gpiomem0 is an alias, so it's file length returns 0
+	fileFlags := os.O_RDWR | os.O_SYNC
+	b.memFile, err = os.OpenFile("/dev/gpiomem0", fileFlags, 0666) // 0666 is an octal representation of: file is readable / writeable by anyone
+	if err != nil {
+		return fmt.Errorf("failed to open %s: %w\n", memPath, err)
+	}
+
+	pageSize := uint64(syscall.Getpagesize())
+	pageOffset := b.physAddr & (pageSize - 1) // difference between base address of the page and the address we're actually tring to access
+	//pageStartAddr := int64(b.physAddr - pageOffset) // base address of the page. needed when we can't open dev/gpiomem0
+	lenMapping := int(pageOffset) + int(b.chipSize) // total amount of memory needed to be mapped. + pageoffset is because we're starting from the base address of the page, not our actual physical address
+
+	vPage, err := mmap.MapRegion(b.memFile, lenMapping, mmap.RDWR, 0, 0) // 0 flag = shared, 0 offset because we are starting from base address pointing to /gpiomem0
+	if err != nil {
+		if err == syscall.ENOMEM {
+			return fmt.Errorf("failed to mmap: cannot allocate memory for mmap call %w", err)
+		}
+		return fmt.Errorf("failed to mmap: %w\n", err)
```

Comment:
when you wrap the error here are you able to see the syscall error code?

Reply:
Not exactly sure what you mean by this - I can return the error code in the message though. Also, this reminds me that I should change the err code to equal the equivalent error in the mmap library, since we learned that the flags are different.